### PR TITLE
Implement Temporal metrics collector

### LINKED ISSUE #16: Temporal metric collector
We should support Temporal as a metrics endpoint via their Go SDK: https://github.com/temporalio/sdk-go

### PR DESCRIPTION
**Pull request by benbjohnson**
This pull request implements a metrics collection client for [Temporal](https://temporal.io/). It uses the `ListWorkflow()` endpoint to fetch a list of `Running` workflow executions to determine the "queue depth". This approach has its limitations as it does not account for activities.

An alternative is to use the Prometheus metrics collector to read the `temporal_worker_task_slots_available` metric.

See also the KEDA Temporal scaler which was closed: https://github.com/kedacore/keda/pull/4863

Fixes #16 

## Usage

To use the Temporal metrics collector, you can specify the config as environment variables:

```
FAS_TEMPORAL_METRIC_NAME="my_workflow_count"
FAS_TEMPORAL_HOSTPORT="localhost:7233"
FAS_TEMPORAL_CERT_DATA="..."
FAS_TEMPORAL_KEY_DATA="..."
FAS_TEMPORAL_QUERY='WorkflowType="my_workflow_type"'
```

Or you can specify it in the `fly-autoscaler.yml` config:

```yml
metric-collectors:
  - type: "temporal"
    metric-name: "workflow_count"
    address: "localhost:7233"
    cert-data: "-----BEGIN CERTIFICATE-----..."
    key-data: "-----BEGIN EC PRIVATE KEY-----..."
    query: 'WorkflowType="my_workflow_type"'
```

## TODO

- [x] Wire up configuration in the `fly-autoscaler` binary.
- [x] End-to-end testing.

### CODE REVIEW

File: temporal/temporal.go [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+	if err != nil {
+		return 0, err
+	}
+	return float64(n), nil
+}
+
+func (c *MetricCollector) workflowExecutionN(ctx context.Context, query string) (n int, err error) {
+	var nextPageToken []byte
+	for {
+		resp, err := c.client.ListWorkflow(ctx, &workflowservice.ListWorkflowExecutionsRequest{
```
Comment (cretz) [inline temporal/temporal.go:86]: At Temporal we don't recommend this approach to scaling (list is eventually consistent and this doesn't help activities in any way), but if you must use workflow count to scale, consider using `CountWorkflow` instead.

Comment (benbjohnson) [reply temporal/temporal.go:86]: Thanks, @cretz. I'll take a look at that worker performance guide. I saw the discussion on the KEDA scaler PR and the trade-offs make sense. This implementation was based on one of our customer's existing implementations so I was trying to stick to that at first.

I saw the `CountWorkflow()` call but IIRC it said somewhere that it was only available if you had Elastic set up. I'm just testing with the local Temporal dev server so I figured it wasn't available without a more extensive setup.

Comment (cretz) [reply temporal/temporal.go:86]: > This implementation was based on one of our customer's existing implementations so I was trying to stick to that at first.

This won't properly scale for activities.

> I saw the CountWorkflow() call but IIRC it said somewhere that it was only available if you had Elastic set up

In all modern versions of the server, this works regardless of whether you have "advanced visibility" (i.e. Elastic) setup. It should work with any local Temporal setup too.

Comment (benbjohnson) [reply temporal/temporal.go]: > This won't properly scale for activities.

I'm ok with that for this first cut of the scaler. The user for the original use case has relatively simple workflows so this approach works for them. We can tweak this and improve it in a separate PR.

> It should work with any local Temporal setup too. If not, can you describe which version you're using?

I switched up the code to use `CountWorkflow()` and the basic test seems to run well. Thanks!

**Comment by hiramhuang** [comment]
Hi @benbjohnson,

I'm looking for a simple solution to auto-scale our Temporal workers, and I found this PR. I'm looking forward to deploying our workers on Fly if this can work.

My question is, can I refer to the "current number of started machines" in the `started-machine-count` expression? I'm planning to start new machines incrementally when the `temporal_worker_task_slots_available` is running low.

```yml
started-machine-count: "min_available_slots > required_min_slots ? started_machines : started_machines + 2"

metric-collectors:
  - type: "prometheus"
    metric-name: "min_available_slots"
    address: "https://api.fly.io/prometheus/MY_ORG"
    query: "min(temporal_worker_task_slots_available)"
    token: "FlyV1 ..."
```

I've tried to look this up first, but it seems that currently only metrics are passed to the Expr compiler as variables.

https://github.com/superfly/fly-autoscaler/blob/5125fb6d50035164fcaa5be4cdb1072f11cf3df8/reconciler.go#L418-L449

**Pull request merged**